idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt

```diff
@@ -92,6 +112,36 @@ internal class KtFirExpressionTypeProvider(
         else -> null
     }
 
+    private fun getExpectedTypeByVariableAssignment(expression: PsiElement): KtType? {
+        // Given: `x = expression`
+        // Expected type of `expression` is type of `x`
+        val assignmentExpression =
+            expression.unwrapQualified<KtBinaryExpression> { binaryExpr, expr -> binaryExpr.right == expr && binaryExpr.operationToken == KtTokens.EQ }
+                ?: return null
+        val variableExpression = assignmentExpression.left as? KtNameReferenceExpression ?: return null
+        return getKtExpressionType(variableExpression)
+    }
+
+    private fun getExpectedTypeByPropertyDeclaration(expression: PsiElement): KtType? {
+        // Given: `val x: T = expression`
+        // Expected type of `expression` is `T`
+        val property = expression.unwrapQualified<KtProperty> { property, expr -> property.initializer == expr } ?: return null
+        val propertyReturnType = getReturnTypeForKtDeclaration(property)
+
+        // Property return type can be a KtErrorType if there is no explicit type in property declaration AND the expression is unresolved.
+        return propertyReturnType.takeIf { it !is KtErrorType }
+    }
+
+    private fun getExpectedTypeByFunctionExpressionBody(expression: PsiElement): KtType? {
+        // Given: `fun f(): T = expression`
+        // Expected type of `expression` is `T`
+        val function = expression.unwrapQualified<KtFunction> { function, expr -> function.bodyExpression == expr } ?: return null
+        val functionReturnType = getReturnTypeForKtDeclaration(function)
+
+        // Function return type can be a KtErrorType if there is no explicit type in property declaration AND the expression is unresolved.
+        return functionReturnType.takeIf { it !is KtErrorType }
```

Comment:
Let's move that check to `getExpectedType` because it is good idea to not return error type from `getExpectedType` at all, and also specify that contract in KtExpressionTypeProviderMixIn.getReturnKtType KDoc